### PR TITLE
[Concurrency] Suggest adding a method, when mutating actor property cross isolation

### PR DESCRIPTION
**Pull request by ktoso**
This comes as an idea from the [Why do not actor-isolated properties support 'await' setter](https://forums.swift.org/t/why-do-not-actor-isolated-properties-support-await-setter/73867/26) thread.

The note that
> Actor-isolated property 'colorFill' can not be mutated from the main actor

Leads people to assume that this is somehow "broken" while they could add a metho to perform the mutation and move on.

A developer comments that:

> So I just assumed that this meant that it was fundamentally wrong to
> mutate this actor's properties from another actor, while all I actually
> needed was... a 3 line setter.

resolves rdar://134041422

### CODE REVIEW

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -11701,51 +11705,29 @@ ActorIsolation ActorIsolation::forActorInstanceSelf(NominalTypeDecl *selfDecl) {
...
-    return actorType
-        ->getReferenceStorageReferent()->getAnyActor();
+  if (getKind() == GlobalActor) {
+    return getGlobalActor()->getAnyNominal();
   }
 
-  return actorInstance.get<NominalTypeDecl *>();
-}
-
-NominalTypeDecl *ActorIsolation::getActorOrNullPtr() const {
```
Comment (ktoso) [inline lib/AST/Decl.cpp:11729]: You added this one recently but it seems to be a 99% copy paste of the other and it is not used so I removed it @gottesmm if we should keep it let me know.

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -11701,51 +11705,29 @@ ActorIsolation ActorIsolation::forActorInstanceSelf(NominalTypeDecl *selfDecl) {
...
-  }
-
-  if (actorType) {
-    if (auto wrapped = actorType->getOptionalObjectType()) {
-      actorType = wrapped;
-    }
-    return actorType
-        ->getReferenceStorageReferent()->getAnyActor();
+  if (getKind() == GlobalActor) {
+    return getGlobalActor()->getAnyNominal();
```
Comment (ktoso) [inline lib/AST/Decl.cpp:11714]: `ActorIsolation::getActor()` used to return garbage before this fix.

We initialize the `globalActor` property but fall through to returning the `actorInstance.get<NominalTypeDecl *>();` which is just some uninitialized memory for the global actor case.

This is fixed now, we properly handle both kinds of isolations the method claimed to support.

Just FYI @hborla

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -1823,6 +1823,52 @@ static bool wasLegacyEscapingUseRestriction(AbstractFunctionDecl *fn) {
...
+
+  if (!refCxt->isAsyncContext()) {
+      // don't suggest creating method when in sync context, as we won't be able
+      // to invoke it anyway so this would not be helpful to suggest
+      return;
+  }
+
+  if (auto actor = isolation.getActor()) {
+      // TODO: isolation is printing without the type name for instance actors,
+      //  it'd be nicer to print with type name and avoid this switch
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:1857]: The ways the isolation is printed ended up not work with the sentence too well; we always want to offer the actor name, while `actorIsolation` printing did not print that for instance isolation, so for now this workaround.

File: test/Concurrency/Inputs/GlobalVariables.swift [hunk 4]
```diff
@@ -2,7 +2,7 @@ public struct Globals {
   public static let integerConstant = 0
   public static var integerMutable = 0
 
-  public static nonisolated(unsafe) let nonisolatedUnsafeIntegerConstant = 0
```
Comment (ktoso) [inline test/Concurrency/Inputs/GlobalVariables.swift:5]: this caused an un-necessary warning we're not testing for; removed the warning

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 5]
```diff
@@ -5455,6 +5455,12 @@ ERROR(distributed_actor_isolated_non_self_reference,none,
       "distributed actor-isolated %kind0 can not be accessed from a "
       "nonisolated context",
       (const ValueDecl *))
+NOTE(note_consider_method_for_isolated_property_mutation,none,
+     "consider declaring an %0 %1 method to perform the mutation",
+     (const NominalTypeDecl *, ActorIsolation))
+NOTE(note_consider_method_for_global_actor_isolated_property_mutation,none, // TODO: de-duplicate with note_consider_method_for_isolated_property_mutation
```
Comment (hborla) [inline include/swift/AST/DiagnosticsSema.def:5461]: How about phrasing this as

```
consider declaring an isolated method on %0 to perform the mutation
```

where `%0` is just a `Type`, which works for both kinds of actor isolation?

Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:5461]: That sounds good :) easier to make that diagnostic too, thanks

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 6]
```diff
@@ -1823,6 +1823,54 @@ static bool wasLegacyEscapingUseRestriction(AbstractFunctionDecl *fn) {
...
+      NominalTypeDecl *actorTy =
+        actor;
+//          (isolation.getKind() == swift::ActorIsolation::ActorInstance ?
+//)
+
+      C.Diags.diagnose(
+          memberLoc,
+          diag::note_consider_method_for_isolated_property_mutation,
+          actor);
+//      } else if (isolation.getKind() == swift::ActorIsolation::GlobalActor) {
```
Comment (tshortli) [inline lib/Sema/TypeCheckConcurrency.cpp:1865]: Was the commented out code here and above meant to be left in? `NominalTypeDecl *actorTy` is also unused

**Pull request merged**